models/classes/class.TestsService.php

```diff
@@ -344,31 +338,49 @@ public function setTestModel(core_kernel_classes_Resource $test, core_kernel_cla
 					$former->deleteContent($test);
 				}	
 			}
-			$test->editPropertyValues(new core_kernel_classes_Property(self::PROPERTY_TEST_TESTMODEL), $testModel);
+			$test->editPropertyValues($this->getProperty(self::PROPERTY_TEST_TESTMODEL), $testModel);
 			$newImpl = $this->getTestModelImplementation($testModel);
 			if (!empty($newImpl)) {
 				$newImpl->prepareContent($test, $items);
 			}
 		}
     }
 
+    /**
+     * Returns a compiler instance for a given test
+     * @param core_kernel_classes_Resource $test
+     * @param ServiceFileStorage $storage
+     * @return tao_models_classes_Compiler
+     */
+    public function getCompiler(core_kernel_classes_Resource $test, ServiceFileStorage $storage) {
+        $testModel = $this->getTestModelImplementation($this->getTestModel($test));
+        if ($testModel instanceof TestModel) {
+            $compiler = $testModel->getCompiler($test, $storage);
+        } else {
+            $testCompilerClass = $testModel->getCompilerClass();
+            $compiler = new $testCompilerClass($test, $storage);
+            $compiler->setServiceLocator($storage->getServiceLocator());
+        }
+        return $compiler;
+    }
+
     public function getCompilerClass(core_kernel_classes_Resource $test) {
         $testModel = $this->getTestModel($test);
-        if (is_null($testModel)) {
-            throw new common_exception_Error('undefined testmodel for test '.$test->getUri());
-        }
         return $this->getTestModelImplementation($testModel)->getCompilerClass();
     }
     
     /**
      * Returns the model of the current test
      * 
      * @param core_kernel_classes_Resource $test
-     * @return core_kernel_classes_Container
+     * @return core_kernel_classes_Resource
      */
     public function getTestModel(core_kernel_classes_Resource $test) {
 		$testModel = $test->getOnePropertyValue(new core_kernel_classes_Property(self::PROPERTY_TEST_TESTMODEL));
-		return $testModel instanceof core_kernel_classes_Resource ? $testModel : null;
+		if (is_null($testModel)) {
+		    throw new common_exception_Error('undefined testmodel for test '.$test->getUri());
```

Comment:
Will something catch this? Because returning null vs throwing an exception is totally different behavior.

Reply:
The exception was just moved up from getCompilerClass to getTestModel so we do not have to throw the same exception from multiple locations

Reply:
yes I see it, but getTestModel is called from several places

Reply:
In general this improves the error handling, but there is 1-2 places that could be negatively impacted, I will double check

Reply:
Added dedicated excaption and fixed error handling